File: 算法/代码随想录算法PDF/00调试源代码/239.滑动窗口最大值.cpp
/*
 * @lc app=leetcode.cn id=239 lang=cpp
 *
 * [239] 滑动窗口最大值
 */

// @lc code=start
class MyQueue {  // 单调队列
    public:
        deque<int> que;  // 底层用双向队列实现

        void push(int x) {  // 直到末尾元素大于x或全部pop，才将元素放入队尾
            while (!que.empty() && que.back() < x) {
                que.pop_back();
            }
            que.push_back(x);
        }

        void pop(int x) {  // 只有队头元素等于x，才把它pop出来，否则不动
            if (!que.empty() && que.front() == x) {
                que.pop_front();
            }
        }

        int maxVal() {  // 获取最大元素（当前队头元素）
            return que.front();
        }
};
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        MyQueue que;
        vector<int> res;
        for (int i = 0; i < k; i++) {
            que.push(nums[i]);
        }
        res.push_back(que.maxVal());
        for (int i = k; i < nums.size(); i++) {
            que.pop(nums[i - k]);
            que.push(nums[i]);
            res.push_back(que.maxVal());
        }
        return res;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/239.滑动窗口最大值.py
#
# @lc app=leetcode.cn id=239 lang=python3
#
# [239] 滑动窗口最大值
#

# @lc code=start
from collections import *
class MyQueue:
    # 初始化
    def __init__(self) -> None:
        self.que = deque()

    # push新元素，把小于它的全清除
    def push(self, x) -> None:
        while self.que and self.que[-1] < x:
            self.que.pop()
        self.que.append(x)
    
    # pop头部元素，只在相等时pop
    def pop(self, x) -> None:
        if self.que and self.que[0] == x:
            self.que.popleft()
    
    # 获取最大元素
    def maxVal(self) -> int:
        return self.que[0]

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        que = MyQueue()
        res = []
        for i in range(k):
            que.push(nums[i])
        res.append(que.maxVal())
        for i in range(k, len(nums)):
            que.pop(nums[i - k])
            que.push(nums[i])
            res.append(que.maxVal())
        return res
# @lc code=end

